Issue 5703 - can't have unnest and window aggregate

Tmonster opened the pull request:
This should provide more error information for https://github.com/duckdb/duckdb/issues/5703.

We don't want to allow window aggregates and unnest clauses in the same select statement since the window is planned before an unnest, but we need to unnest first to plan the window aggregate.

Tmonster committed:
cannot have unnest in a window statement

Tmonster committed:
mistype

Tmonster committed:
check window aggregate function

Tmonster committed:
small improvements

Mytherin commented on test/issues/general/test_5703.test:
```diff
@@ -0,0 +1,19 @@
+----
+Binder Error
+
+# check database is not invalidated
```
This is not necessary - an `InternalError` always fails a test (even if the test is labeled as `statement error`)

Mytherin commented on src/planner/binder/query_node/bind_select_node.cpp:
```diff
@@ -445,6 +445,7 @@ unique_ptr<BoundQueryNode> Binder::BindNode(SelectNode &statement) {
 	vector<LogicalType> internal_sql_types;
 	for (idx_t i = 0; i < statement.select_list.size(); i++) {
 		bool is_window = statement.select_list[i]->IsWindow();
+		bool is_window_aggregate = statement.select_list[i]->type == ExpressionType::WINDOW_AGGREGATE;
```
This checks that the window function is specifically a window aggregate - it misses other window functions (such as `WINDOW_LEAD`, e.g. `lead(r) over ()`):

```sql
select unnest([5,4,3]) as r, lead(r) OVER ();
```

Could you add a test for that as well to catch this case?

Mytherin replied:
This also only checks that the root element is a window clause, what if we have something like this:

```sql
select unnest([5,4,3]) as r, sum(r) OVER () + 42;
```

Mytherin commented on src/planner/binder/query_node/bind_select_node.cpp:
```diff
@@ -461,6 +462,9 @@ unique_ptr<BoundQueryNode> Binder::BindNode(SelectNode &statement) {
 			result->groups.group_expressions.push_back(std::move(expr));
 			expr = std::move(group_ref);
 		}
+		if (!result->unnests.empty() && is_window && is_window_aggregate) {
```
`result->unnests` contains the unnests for the entire query - this means that unnests and window functions cannot be mixed. For example, will this query still work:

```sql
select unnest([5, 4, 3]), sum(41) over ();
```

Mytherin reviewed (commented):
Thanks for the PR! Looks good - some comments below:

Tmonster committed:
add more tests

Tmonster committed:
error is thrown in planning stage now, and does a better job of checking the order of the operators

Tmonster committed:
Merge remote-tracking branch 'upstream/master' into issue-5703-throw-better-error

Tmonster commented:
@Mytherin All very good points that forgot to think about. I've put the check in the planning phase now, to make sure that an unnest can't be above a window operator, which should avoid throwing errors for unnest and window operations that are not stacked on top of each other. I can imagine that I am still missing something, or that I'm throwing an incorrect error.

Tmonster committed:
make format fix

Mytherin commented on test/issues/general/test_5703.test:
```diff
@@ -0,0 +1,63 @@
+Binder Error
+
+statement error
+select unnest([5,4,3]) as r, sum(r) OVER () + 42;
```
What happens if you have the following query:

```sql
select unnest([5,4,3]) as r, sum(r + 42) OVER ();
```

Tmonster replied:
This also fails. Have added a test

Mytherin commented on src/planner/binder/query_node/plan_select_node.cpp:
```diff
@@ -96,6 +95,14 @@ unique_ptr<LogicalOperator> Binder::CreatePlan(BoundSelectNode &statement) {
+			for (auto &expression : root->expressions) {
+				if (expression->type != ExpressionType::WINDOW_ROW_NUMBER) {
+					throw BinderException(
+					    "Cannot have window expression over an unlist unless the window funciton is row_number()");
```
I assume the reason `row_number` is exempt is because it does not have any arguments, which causes no issues to arise? Would it not be possible to instead look at the actual arguments to check if an issue would arise, rather than making a specialization on row number here?

Tmonster replied:
Yea, row number is exempt for that reason.

I can check the arguments to make sure nothing is bound to the unnest alias. Is there a clean way to get the alias string of the unnest? I have access to the unnest object, but for some reason the alias for the unnest expression is not initialized.

There are other problems with this though. In the last test I have 
`SELECT unlist(str_split('A',' ')) AS "r", CASE WHEN ("r" = 'A') THEN (-1) ELSE 0 END AS "x", sum("x") OVER () AS "s";`

the window is over an alias to an alias of an unlist, so I would also have to check for that case.

Also, when I call `EnumerateChildren` to check the argument types, the function doesn't then recursively enumerate the children. Is this intended behavior? If I have a the test case  `select unnest([5,4,3]) as r, sum(r + 42) OVER ();` it will pass since the children of the `bound_function` expression are not checked. 

I can probably find ways to solve each of the above issues, but I think just prohibiting it in the current way is cleaner. Users can still put unnest statements in the `from` clause

Mytherin replied:
It might be easier to look at the number of elements in the `BoundSelectNode::unnests` list rather than looking at the children. If the number increases then `UNNEST` clauses have been bound.

Tmonster replied:
So I don't think the number of children in the unnest increases when binding the window because the unnest clause comes first in the select list and is therefore bound first. However, during the window binding I can check what ColumnRef alias the window is bound to. 

I added some functionality to check for that and I think the current solution is a bit more elegant.

Mytherin reviewed (commented):
Thanks for the updates - some more comments

Tmonster committed:
add test for function inside the sum aggregate

Tmonster committed:
maybe check the arguments of the window function to make sure the window doesn't use unnest variables

Tmonster committed:
revert some changes from when I was exploring ideas

Tmonster committed:
Merge remote-tracking branch 'upstream/master' into issue-5703-throw-better-error

Tmonster committed:
this fix makes more sense. Throw error when binding a window and the child is an unnest

Tmonster committed:
make format-fix

Tmonster committed:
Merge remote-tracking branch 'upstream/master' into issue-5703-throw-better-error

Mytherin reviewed (commented):
Thanks for the updates - some more comments

Tmonster committed:
reverting unnest bool attribute from boundcolref

Tmonster committed:
now the select binder keeps track of which nodes in the select statement are unnest nodes. Any select nodes that are unnests are also added to the unnest indexes.

Tmonster committed:
Added some more tests and functionality'

Tmonster committed:
some more code clean up

Tmonster commented:
@Mytherin so I looked into it again. I don't really want to touch the alias_map, because mutliple binders seem to have an alias map, and I don't want just the select binder alias map to behave differently because of this bug.
Right now I've added an unnest_indexes vector (realizing now it can just be a set) to the select binder. When an unnest function expression is bound, the unnest_indexes keeps track of where in the select list the unnest is. Each time BindColumnRef is called, we can check if we are in a window, and if the index of the aliased node is an unnest or not.

Also added functionality to check if a child of a window is an unnest.

This doesn't seem like a perfect solution either though. I think at this point it might be both if we both sit down and look at the issue together. The problem is once the unnest function expression node is bound, it becomes a BoundColumnRef and I don't think there is a way to keep track of its original expression type.

Tmonster committed:
add unnest function and function_expression.hpp to avoid incomplete type error

Tmonster committed:
make format-fix

Tmonster committed:
Merge remote-tracking branch 'upstream/master' into issue-5703-throw-better-error
